mmd_tools/core/model.py

```diff
@@ -633,189 +642,133 @@ def join_models(parent_root_object: bpy.types.Object, child_root_objects: Iterab
 
                     if temp_objects:
-                        # Ensure we're in object mode
-                        if bpy.context.mode != "OBJECT":
-                            bpy.ops.object.mode_set(mode="OBJECT")
+                        # Use parent_set with keep_transform=True for temporary objects
+                        with select_object(parent_temporary_group_object, objects=[parent_temporary_group_object] + temp_objects):
+                            bpy.ops.object.parent_set(type="OBJECT", keep_transform=True)
 
-                        for temp_obj in temp_objects:
-                            # Save world coordinate position
-                            original_matrix_world = temp_obj.matrix_world.copy()
+                    # Remove child objects and groups
+                    child_objects = list(FnModel.iterate_child_objects(child_temporary_group_object))
+                    for obj in child_objects:
+                        bpy.data.objects.remove(obj)
+                    bpy.data.objects.remove(child_temporary_group_object)
 
-                            # Set parent object
-                            temp_obj.parent = parent_temporary_group_object
-                            temp_obj.parent_type = "OBJECT"
+            # Copy MMD root properties
+            FnModel.copy_mmd_root(parent_root_object, child_root_object, overwrite=False)
 
-                            # Restore world coordinate position
-                            temp_obj.matrix_world = original_matrix_world
+        # Clean additional transform before join
+        bpy.ops.object.mode_set(mode="OBJECT")
+        FnContext.set_active_and_select_single_object(bpy.context, parent_root_object)
+        bpy.ops.mmd_tools.clean_additional_transform()
+        for child_root_object in child_root_objects:
+            FnContext.set_active_and_select_single_object(bpy.context, child_root_object)
+            bpy.ops.mmd_tools.clean_additional_transform()
+
+        # Join all child armatures to parent armature
+        bpy.ops.object.mode_set(mode="OBJECT")
+        child_armature_objects = [FnModel.find_armature_object(child_root) for child_root in child_root_objects if FnModel.find_armature_object(child_root) is not None]
+        armature_data_to_remove = [child_arm.data for child_arm in child_armature_objects if child_arm.data]
+        if child_armature_objects:
+            with select_object(parent_armature_object, objects=[parent_armature_object] + child_armature_objects):
+                bpy.ops.object.join()
+        for armature_data in armature_data_to_remove:
+            if armature_data and armature_data.users == 0:
+                bpy.data.armatures.remove(armature_data)
+
+        # Apply additional transform after join
+        bpy.context.view_layer.objects.active = parent_root_object
+        bpy.ops.mmd_tools.clean_additional_transform()
```

Comment:
I suggest using `FnBone.clean_additional_transformation` instead of `bpy.ops.mmd_tools.clean_additional_transform`.

In general, it is best to avoid operator calls.
Because operator calls are slow and can have unexpected side effects.

Reply:
Understood.
As a long-term task, we need to organize the core functions.